spec/requests/admin_payroll_run_downloads_spec.rb

```diff
@@ -1,78 +1,129 @@
+            payroll_run = create(:payroll_run, downloaded_at: Time.zone.now, downloaded_by: admin_session_id)
 
-    get admin_payroll_run_download_path(payroll_run)
+            get admin_payroll_run_download_path(payroll_run)
 
-    expect(response.body).to include "Download the payroll run file for #{Date.today.strftime("%B")}."
-    expect(response.body).to include admin_payroll_run_download_path(payroll_run, format: :csv)
-  end
+            expect(response.body).to include admin_payroll_run_download_path(payroll_run, format: :csv)
 
-  it "shows when and which user download the payroll run file when payroll run has been marked as downloaded" do
-    payroll_run = create(:payroll_run, downloaded_at: Time.zone.now, downloaded_by: "admin_user_id")
+            travel_to 31.seconds.from_now do
+              get admin_payroll_run_download_path(payroll_run)
 
-    travel_to 31.seconds.from_now do
-      get admin_payroll_run_download_path(payroll_run)
+              expect(response.body).not_to include admin_payroll_run_download_path(payroll_run, format: :csv)
+            end
+          end
+        end
 
-      expect(response.body).to include payroll_run.downloaded_by.to_s
-      expect(response.body).to include I18n.l(payroll_run.downloaded_at)
-    end
-  end
-end
+        context "and the timeout has been reached" do
+          it "shows who triggered the download and when" do
+            payroll_run = create(:payroll_run, downloaded_at: 31.seconds.ago, downloaded_by: "admin_user_id")
 
-  describe "downloads#create" do
-    let(:admin_user_id) { "admin_user_id" }
+            get admin_payroll_run_download_path(payroll_run)
 
-    it "sets the downloaded_at and downloaded_by attributes on the payroll_run" do
-      sign_in_to_admin_with_role(AdminSession::SERVICE_OPERATOR_DFE_SIGN_IN_ROLE_CODE, admin_user_id)
+            expect(response.body).to include payroll_run.downloaded_by
+            expect(response.body).to include I18n.l(payroll_run.downloaded_at)
+          end
+        end
+      end
 
-      payroll_run = create(:payroll_run)
-      downloaded_at = Time.zone.now
+      context "when requesting csv" do
+        context "and it is within the timeout" do
+          it "allows the payroll run file to be downloaded within the time limit" do
+            payroll_run = create(:payroll_run, downloaded_at: Time.zone.now, downloaded_by: "admin_user_id")
+            get admin_payroll_run_download_path(payroll_run, format: :csv)
 
-      travel_to downloaded_at do
-        post admin_payroll_run_download_path(payroll_run)
+            expect(response.headers["Content-Type"]).to eq("text/csv")
+          end
+        end
 
-        expect(payroll_run.reload.downloaded_by).to eql admin_user_id
-        expect(payroll_run.downloaded_at.to_s).to eql downloaded_at.to_s
+        context "and the timeout has been reached" do
+          it "redirects a request for the file once the timeout has been reached" do
```

Comment:
I think that there are some changes to the structure of this file, not related to the authentication, that should probably be moved to the previous commit? Looks like some of these changes might address some comments I made on the previous commit.

Reply:
I'm still struggling to read the diff of this commit, titled "Payroll operators can download payroll run file", for the reasons mentioned before. It looks like the main changes to the specs are a re-indentation so that you can have the `context "when signed in as a payroll operator" do` block, and the addition of the "when signed in as a service operator" block. But then, there are other changes that seem like they belong in the previous commit, like `[:html, :csv].each do |format|` or `context "when requesting html" do`. Do you think you'd be able to move these changes into the appropriate commit?

Reply:
Yeah, I've done something to break this. I'll try and sort it in the morning. Thanks for being patient.

Reply:
I am very much hoping this is all sorted now.